gl: fix corruption when initial data is prepared on multiple contexts

corngood opened the pull request:
If a resource was marked dirty on multiple contexts, it could be enqueued by
Prepare_InitialState multiple times, and each successive time would delete the
buffer previously referenced in the queue.

- don't reallocate initial data if it exists
- get data pointer from resource manager instead of storing it in the queue
- early out from ContextPrepare_InitialState if valid data exists
- only prepare resources for current context in ActivateContext

corngood commented:
WIP due to lack of testing, and a couple of concerns I had with naming etc.

corngood commented on renderdoc/driver/gl/gl_initstate.cpp:
```diff
@@ -484,10 +498,14 @@ bool GLResourceManager::Prepare_InitialState(GLResource res)
   {
-    byte *data = AllocAlignedBuffer(sizeof(FramebufferInitialData));
-    RDCEraseMem(data, sizeof(FramebufferInitialData));
+    if(!GetInitialContents(Id).blob)
```
Maybe `HasInitialContents`, or add operator == and test against `InitialContentData()` ?

corngood commented:
What I'm seeing is a resource (In my case a VBO) gets added to the dirty list on multiple contexts, but each time it queues up the prepare, it deletes the buffer from the previous prepare, causing memory corruption.

I also found that ActivateContext could prepare resources from other contexts, and it didn't look like that was the intention.

corngood commented:
... actually I may have misunderstood why the buffers were being deleted.  I'll do a bit more investigation.

baldurk commented:
How is this situation getting hit - are you rendering on multiple contexts simultaneously? That use case isn't supported and there's a whole bunch of other things that won't work beyond initial states. If it's context switching on a single thread then that's what the queued initial state preparation is intended to solve.

corngood commented:
Yeah, it's just context switching on a single thread.

baldurk commented:
OK, I'd expect that to work then. In which case I need more information about the sequence of events to figure out what's going wrong. Some of these changes I think are good (not overrunning the range in the `m_QueuedInitialFetches` loop, renaming the overload to `ContextPrepare_InitialState` to clarify).

For buffers, since they're shared between contexts it should be the easy case. We should be able to fetch their contents as long as any context is current at the time we start. The only time we'd need to enqueue initial state fetch for buffers is theoretically if there was no context current at all, at which point we'd fetch them the first time one is activated. Though I'm not sure I even correctly handle that case.

Marking a resource as dirty twice doesn't track with how things should work. A resource is either dirty or not, we shouldn't care how many times it's marked. While we're background capturing we note which resources we're going to need initial states for, then once we are doing the active capture we prepare and serialise each one, then don't mark anything dirty until we've stopped the frame capture.

corngood commented:
So far what I've found is that there end up being two entries in the dirty list for the same GLResource (ns=VA, name=1 in my case) for a single context.  The two entries have different ResourceIDs, and I can see (via logs) the RegisterResource calls for those objects.

It doesn't crash 100% for me, and so far I've only reproduced it after 2 or more captures from the same process.

I haven't tracked it any further than that yet (current test project is rather unwieldy).

Does it make sense for the same object to be registered twice?

baldurk commented:
For VAOs it's possible because they shouldn't be shared between contexts, so you could have (ns=VAO, name=1, ctx=0x1000) and (ns=VAO, name=1, ctx=0x1001). They should be treated as distinct objects, and this is where the queuing of initial states will come in; if 0x1000 is active when the capture starts then the first initial state will be prepared immediately, the second will be queued for if/when 0x1001 is activated.

If the ctx is NULL then that means the vendor check for VAO sharing has tested that VAOs *are* shared between contexts. We then expect that the implementation will not return name=1 from glGenVertexArrays while name=1 is still around on some other context. There's a log message "VAOs are shared on this implementation" and similarly for FBOs. If the GL implementation flip-flopped and shared for the test but then didn't share later, that would explain it.

Also check that all of your contexts are set to share with each other. I assume in the code that all contexts are shared with all others, so if you have distinct share groups of contexts then that will break things and possibly lead to this case (e.g. if the vendor check indicates sharing so VAOs are created with ctx=0, but then there are non-shared contexts so we get name=1 returned on two different contexts as two different objects).

corngood commented:
I'm actually getting duplicates in the same context, at least in `m_CurrentResourceMap` (see below).  That seems bad, right?

I'm just stepping through now to try to find out how these VAOs get registered/unregistered.

```
(gdb) p id.id = 2648
$5 = 2648
(gdb) p *m_CurrentResourceMap.find(id)
$8 = (std::pair<ResourceId const, GLResource> &) @0x31d77eb0ef20: {
  first = {
    id = 2648
  }, 
  second = {
    Context = 0x31d77e824a80, 
    Namespace = eResVertexArray, 
    name = 1
  }
}
(gdb) p id.id = 2687
$9 = 2687
(gdb) p *m_CurrentResourceMap.find(id)
$10 = (std::pair<ResourceId const, GLResource> &) @0x31d77e8909a0: {
  first = {
    id = 2687
  }, 
  second = {
    Context = 0x31d77e824a80, 
    Namespace = eResVertexArray, 
    name = 1
  }
}
```

baldurk commented:
That's definitely not good!

The normal flow is that the resource is registered in the loop in either `WrappedOpenGL::glGenVertexArrays` or `WrappedOpenGL::glCreateVertexArrays` depending on which function is called. Then it will be unregistered and removed from that map you're looking at (and others) in `WrappedOpenGL::glDeleteVertexArrays`.

The most likely explanation then is that somehow the VAO is getting destroyed without going through that path `WrappedOpenGL::glDeleteVertexArrays`, so the next VAO that is created returns the same name again, but is given a new ID and the old one still lurks around. How that would happen though I don't exactly know.

corngood commented:
Looks like between captures, contexts are destroyed and created, and the context id gets reused.  It's not actually the same context.  Perhaps it's not cleaning up resource manager in that case?  I don't see anything obvious that would do it.  I'll look into that in more detail.

baldurk commented:
Oh that makes sense, you're right. The code is missing the implicit deletion of all objects when the context is destroyed.

It will probably need two separate checks - one for context-specific objects that deletes them when the context is deleted, and one for context-shared objects that deletes them when all current contexts are deleted. There's a `WrappedOpenGL::DeleteContext` callback but this seems like it'd be best to do as a new method in `GLResourceManager` which has the list of objects ready to iterate over.

Thanks again for delving so deep into this!

corngood committed:
gl: Delete context owned resources from resource manager.

This fixes tracking problems when a context id is reused.

corngood committed:
gl: Avoid passing initial data pointers through the queue.

This reduces the possibility of a logic error leading to memory
corruption by using a dangling pointer.

corngood committed:
gl: Fix initial data preparation on wrong context.

corngood commented:
I split it up a bit.

- stopped passing the data pointer through the queue
  - memory corruption should be less likely
  - renamed to ContextPrepare_InitialState because the parameter types are now the same as the other overload
- fixed that loop error
- added some initial handling of context deletion to gl_manager
  - TODO: deal with shared resources (I would probably do full tracking of sharing groups or nothing)
  - TODO: look at m_SyncIDs etc.
  - it shouldn't be any more broken than before

It seems to fix my problem, but I haven't done much general testing yet.

corngood commented on renderdoc/driver/gl/gl_initstate.cpp:
```diff
@@ -380,6 +391,7 @@ bool GLResourceManager::Prepare_InitialState(GLResource res, byte *blob)
   {
     VAOInitialData *data = (VAOInitialData *)blob;
 
+    RDCASSERT(!data->valid);
```
@baldurk I haven't had any major problems with these changes, but I did hit this assert once during a capture, after quite a long session with many captures.

The only thing I can think of is that a context got deleted during a capture.  I'll see if i can get more info if it happens again.

baldurk replied:
Even if there might be something still up, these changes look good now so I'm happy to merge them (I've made a couple of stylistic comments).

Regarding this assert firing, I don't have any immediate ideas. If the context was deleted and recreated during a capture, and if we created a new VAO with the same resource name on the 'new' context we shouldn't go back to preparing it. The list of prepared resources is fixed at the start of the frame capture: Some will be prepared immediately, some will be queued. I could see maybe a false negative, where we go to prepare an old object and get a new ID with no blob associated, but not the other way around.

It might be interesting (depending on how often you can repro) to see if it's being prepared twice, or if we're poking into free'd memory or something. I'm hoping it's not the latter but it'd be better to check.

I wonder if this might be due to a design problem in the GL driver. In all other drivers, the API objects are wrapped so you can go from object handle -> ID directly as the wrapper object contains its ID. Thus even if you get two identical object handles there won't be any ambiguity about their ID. In GL, we have the lookup table from resource to ID, so if the same resource comes around twice there might be edge cases there.

baldurk commented on renderdoc/driver/gl/gl_driver.cpp:
```diff
@@ -1123,12 +1125,18 @@ void WrappedOpenGL::ActivateContext(GLWindowingData winData)
       // For now we assume we'll only get GL commands from a single thread
       QueuedInitialStateFetch fetch;
       fetch.res.Context = winData.ctx;
+      unsigned before = (unsigned)m_QueuedInitialFetches.size();
```
I'd prefer `uint32_t` here. I don't mind `int` when the specific type doesn't really matter like a for loop counter, but generally I stick to the stdint types for any other cases (i.e. avoiding `long`, `unsigned int`, etc).

corngood replied:
I noticed there's already an instance of %zu being used to format size_t, so I switched these to it instead.  It's what I should have done originally, but I was being lazy about figuring out supported format strings.

Let me know if you'd still prefer uint32_t.

baldurk replied:
No that's fine too, I didn't want to go down the rabbit's hole of "here's the custom printf specifiers I have" which honestly I forget about sometimes :smile:.

baldurk commented on renderdoc/driver/gl/gl_manager.h:
```diff
@@ -92,6 +92,26 @@ class GLResourceManager : public ResourceManager<GLResource, GLResource, GLResou
+      if(it->first.Context == context)
+      {
+        ++count;
+        auto res = it->second;
```
`auto` is fine for iterators where the type is tedious to type out, but I prefer to avoid it when the type can be made clear, like here it's just `ResourceId`.

Also ditto the other comment about `unsigned` for the `count` variable here.

corngood commented:
Looks like travis failed with a timeout on the llvm PPA.
